jwt-bearer-auth-yesod/README.lhs

```diff
@@ -175,42 +194,64 @@ Now, you will almost certainly want to actually inspect the fields on the payloa
 decision. The `ClaimsSet` type in the `jose` library ONLY supports the claims that are required by
 the RFC for JWTs, but you can extend them with more fields. The supported way to do that is by
 providing your own data type that contains a `ClaimsSet` and provide a lens to focus on that
-`ClaimsSet`. For an example, see `Web.Auth.Bearer.JWT.Claims`. You will also need a `FromJSON`
-instance (you would also need a `ToJSON` instance if you were doing the signing, but we're only
-going to be validating ones that have already been signed.)
+`ClaimsSet`. You will also need a `FromJSON` instance; you would also need a `ToJSON` instance if
+you were doing the signing, but we're only going to be validating ones that have already been
+signed. Here is an example (that's already provided for you in `Web.Auth.Bearer.JWT.Claims`):
+
+<!--
+the below example should show up in the README but NOT be compiled with the literate code.
+By using CPP instead of commenting out, it will still show up with the right syntax highlighting.
+```haskell
+#if 0
+```
+-->
 
 ```haskell
-data ClaimsWithScope a = ClaimsWithScope [String] a
-  deriving stock (Show, Eq)
-
-instance FromJSON a => FromJSON (ClaimsWithScope a) where
-  parseJSON = withObject "ClaimsWithScope" $ \v ->
-    ClaimsWithScope
-      <$> (concat <$> v .:? "scp")
-      <*> parseJSON (Object (delete "scp" v))
-
-instance ToJSON a => ToJSON (ClaimsWithScope a) where
-  toJSON (ClaimsWithScope claims scp) =
-    let ~(Object o) = toJSON claims
-     in Object $ insert "scp" (toJSON scp) o
-
-instance HasClaimsSet a => HasClaimsSet (ClaimsWithScope a) where
-  claimsSet = myClaimsSet . claimsSet
-    where
-      myClaimsSet = lens getter setter
-      getter (ClaimsWithScope _ claims) = claims
-      setter (ClaimsWithScope scp _) claims = ClaimsWithScope scp claims
+data JWTClaims extra = JWTClaims ClaimsSet extra
+  deriving stock (Eq, Show)
+
+claimsExtra :: Lens (JWTClaims e1) (JWTClaims e2) e1 e2
+claimsExtra = lens (\(JWTClaims _ e) -> e) (\(JWTClaims c _) e -> JWTClaims c e)
+
+instance FromJSON extra => FromJSON (JWTClaims extra) where
+  parseJSON = withObject "JWTClaims" $ \v ->
+    JWTClaims
+      <$> parseJSON (Object v)
+      <*> parseJSON (Object v)
+
+instance ToJSON extra => ToJSON (JWTClaims extra) where
+  toJSON (JWTClaims claims extra) =
+    let ~(Object jsonClaims) = toJSON claims
+        ~(Object jsonExtra) = toJSON extra
+     in Object (jsonExtra `union` jsonClaims)
+
+instance HasClaimsSet (JWTClaims extra) where
+  claimsSet = lens (\(JWTClaims c _) -> c) (\(JWTClaims _ e) c -> JWTClaims c e)
+
+-- | This "extends" the base '@ClaimsSet@' type by adding an additional "scp"
+--   claim. The ToJSON and FromJSON instances put the "scp" field alongside all
+--   the other fields in the object, not using a separate sub-object.
+data ScpClaims = ScpClaims
+  { scp :: [String]
+  }
+  deriving stock (Eq, Show)
```

Comment:
This all is still going away in some future PR, right?

Reply:
removed the copy-pasted definition of `JWTClaims` and its instances, just keeping the definition of `ScpClaims` as the template for "here's how you define your extra claims"

Reply:
Ah that looks better. I just don't think you want the has class or instance or `get` function, it's all just distracting, IMO.

Reply:
took them out